sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -761,6 +762,48 @@ dynamic parameter for the buffer can be registered with all the nodes which
 use the buffer as a parameter. Then a single `dynamic_parameter::update()` call
 will maintain the graphs data dependencies.
 
+===== Whole Graph Update [[whole-graph-update]]
+
+A graph in the executable state can have all of its nodes updated using the
+`command_graph<graph_state::executable>::update(graph)` method. This method
+takes a source graph in the modifiable state and updates the nodes in the target
+executable state graph to reflect any changes made to the nodes in the source
+graph. The characteristics which will be updated are detailed in the section on
+<<executable-graph-update, Executable Graph Update>>.
+
+Both the source and target graphs for the update must satisfy the following
+conditions:
+
+* Both graphs must have been created with the same device and context.
+* Both graphs must be topologically identical. The graphs are considered
+topologically identical when:
+
+** Both graphs have the same number of nodes and edges.
+** Where edges are specified through the `depends_on` property or through event
+dependencies in recorded queue submisions, these must be specified in the same
+order. 
+** Nodes with dependencies on events outside of the graph must have the same
+number of these dependencies.
+** A kernel node's command function must have the same type in both graphs,
+whether that be a lambda function, function object or otherwise.
+** When creating both graphs all graph-related API calls or operations which
+modify the structure of the graph must have been done in the same order, these
+are:
+
+*** Explicit API calls to add nodes to the graph or calls to `make_edge()`.
+*** Nodes added to a graph via recorded queue submissions.
```

Comment:
This seems good, but could we tighten it a bit more by saying something like:

* Both graphs must have the same number of nodes and edges.
* Nodes must be added in the same order in the two graphs.  Nodes may be added via `command_graph::add` or `queue::submit` (for a recorded queue).
* Corresponding nodes in each graph must be kernels that have the same type:
  * When the kernel is defined as a lambda, the lambda must be the same.
  * When the kernel is defined as a named function object, the kernel class must be the same.
  * When the kernel is defined as a plain function, the function must be the same.
* Events and edges dependencies for each node must be added in the same order in the two graphs.  These may be added via the `depends_on` property, the `handler::depends_on` member function, or the `command_graph::make_edge` member function.
* Internal edges must be between corresponding nodes in each graph.
* External input events in the two graphs must correspond.  If a node _A_ in one graph depends on an event that comes from a command outside of its graph, then the corresponding node _B_ in the other graph must also depend on an event that comes from a command outside of its graph.
* External output events in the two graphs must correspond.  If a node _A_ in one graph produces an event that is not an internal edge in its graph, then the corresponding node _B_ in the other graph must also produce an event that is not an internal edge in its graph.

Reply:
> or queue::submit (for a recorded queue).
>
> These may be added via ..., the handler::depends_on member function

Is this too specific? This wording leaves out queue shortcuts that we do support like `queue::parallel_for()` which also specify event dependencies in a different way (parameter rather than `handler::depends_on()`).

> External output events in the two graphs must correspond. If a node A in one graph produces an event that is not an internal edge in its graph, then the corresponding node B in the other graph must also produce an event that is not an internal edge in its graph.

I don't know that I agree with this one, these output events are only created for specific executable graphs and are per execution. The modifiable source graph has no knowledge of these kinds of events and so cannot be said to be producing any events.

Reply:
> Is this too specific? This wording leaves out queue shortcuts that we do support like `queue::parallel_for()` which also specify event dependencies in a different way (parameter rather than `handler::depends_on()`).

We can definitely improve this, but I think we need to define what "same order" means regarding events and edges.  I actually think my proposed wording is not specific enough.  For example, here are some things that should be clarified:

* If one graph uses the `depends_on` property to define edges for node _A_, does the other graph also need to use the `depends_on` property to define edges for node _A_?  Or is it OK to define edges through a different API so long as the edges are the same?

* If both graphs use the same API to define edges (e.g. the `depends_on` property), do the edges need to be listed in the same order in the two `depends_on` constructors, or does this order not matter?

I think it's OK if we require the exact same APIs and the exact same order if that make life easier.  We just need to say that clearly in the description.

Reply:
Yes I think the exact same APIs and the exact same order makes the most sense. 

If we are mandating that then is it acceptable to then just use say `queue::submit()` as one example of an API, or do you think we need to specify explicitly every possible API that creates a node when recorded?

Reply:
The important thing is that the description must clearly describe what "same order" means.  If you can do this using more general language, that's fine.

Reply:
I've updated this wording based on @gmlueck 's original suggestion and to define edge ordering I've linked to our existing definition of all the ways edges are created and stated that edges "must be created in the same order and using the same API invocation to create each edge". Hopefully that's clear enough to imply "exact same APIs and the exact same order" but happy to refine the wording more with other suggestions.